File: PerfectSquares.java
/**
 * LeetCode 279 Perfect Squares
 */
class Solution {
    public int numSquares(int n) {
        if (n < 0) {
            return -1;
        }
        int[] dp = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            int curMin = Integer.MAX_VALUE;
            for (int j = 1; j * j <= i; j++) {
                curMin = Math.min(curMin,dp[i - (j * j)]);
            }
            dp[i] = curMin + 1;
        }
        return dp[n];
    }
}

File: PerfectSquares.go
func min(a int,b int) int {
	if (a < b) {
		return a
	}
	return b
}

func numSquares(n int) int {
	if n < 2 {
		return n
	}
	dp := make([]int,n + 1)
	dp[0] = 0
	for i := 1;i <= n;i++ {
		dp[i] = math.MaxInt32
		for j := 1;j * j <= i;j++ {
			dp[i] = min(dp[i],dp[i - j * j] + 1)
		}
	}
	return dp[n]
}
